### ISSUE TITLE
numba.cuda.cudadrv.error.NvvmError: Failed to compile

### ISSUE THREAD

**Issue by qfwysw** [open-mmlab/mmdetection3d#513]
Thank you for your outstanding work！
**My Environment**

`fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
sys.platform: linux
Python: 3.7.10 (default, Feb 26 2021, 18:47:35) [GCC 7.3.0]
CUDA available: True
GPU 0,1: GeForce RTX 3090
CUDA_HOME: /usr/local/cuda-11.2
NVCC: Build cuda_11.2.r11.2/compiler.29558016_0
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.7.1+cu110
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.6.0 (Git Hash 5ef631a030a6f73131c77892041042805a06064f)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.0
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80
  - CuDNN 8.0.5
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,

TorchVision: 0.8.2+cu110
OpenCV: 4.5.1
MMCV: 1.2.7
MMCV Compiler: GCC 7.5
MMCV CUDA Compiler: 11.2
MMDetection: 2.10.0
MMDetection3D: 0.12.0+
`
**The command I run**
`python tools/train.py configs/pointpillars/hv_pointpillars_secfpn_6x8_160e_kitti-3d-3class.py`

**Describe the bug.**
`"Import Error in mmdet3d.models.builder", I downgrade mmdet to 2.10.0 and downgrade mmcv to 1.2.7 to solve it. `
`When I start to train, the training looks good. After a while, there is a new error.`
`2021-05-01 15:46:44,930 - mmdet - INFO - Saving checkpoint at 2 epochs
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 47.5 task/s, elapsed: 79s, ETA:     0s
Converting prediction to KITTI format
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 451.3 task/s, elapsed: 8s, ETA:     0s
Result is saved to /tmp/tmp13fakuid/results.pkl.
Traceback (most recent call last):
  File "tools/train.py", line 212, in <module>
    main()
  File "tools/train.py", line 208, in main
    meta=meta)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmdet/apis/train.py", line 170, in train_detector
    runner.run(data_loaders, cfg.workflow)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 125, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 54, in train
    self.call_hook('after_train_epoch')
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmcv/runner/base_runner.py", line 308, in call_hook
    getattr(hook, fn_name)(self)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmdet/core/evaluation/eval_hooks.py", line 147, in after_train_epoch
    key_score = self.evaluate(runner, results)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmdet/core/evaluation/eval_hooks.py", line 177, in evaluate
    results, logger=runner.logger, **self.eval_kwargs)
  File "/home/qf/mm/mmdetection3d/mmdet3d/datasets/kitti_dataset.py", line 355, in evaluate
    self.CLASSES)
  File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 707, in kitti_eval
    eval_types)
  File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 613, in do_eval
    min_overlaps)
  File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 479, in eval_class
    rets = calculate_iou_partly(dt_annos, gt_annos, metric, num_parts)
  File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 382, in calculate_iou_partly
    dt_boxes).astype(np.float64)
  File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 116, in bev_box_overlap
    from .rotate_iou import rotate_iou_gpu_eval
  File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/rotate_iou.py", line 292, in <module>
    criterion=-1):
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/decorators.py", line 101, in kernel_jit
    kernel.bind()
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/compiler.py", line 548, in bind
    self._func.get()
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/compiler.py", line 426, in get
    ptx = self.ptx.get()
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/compiler.py", line 397, in get
    **self._extra_options)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 496, in llvm_to_ptx
    ptx = cu.compile(**opts)
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 233, in compile
    self._try_error(err, 'Failed to compile\n')
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 251, in _try_error
    self.driver.check_error(err, "%s\n%s" % (msg, self.get_log()))
  File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 141, in check_error
    raise exc
numba.cuda.cudadrv.error.NvvmError: Failed to compile

<unnamed> (66, 23): parse expected comma after load's type
NVVM_ERROR_COMPILATION
`

**Comment by Tai-Wang** [comment]
It seems the error is caused by
```
File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 116, in bev_box_overlap
from .rotate_iou import rotate_iou_gpu_eval
```
What's your numba version? Maybe you can try with numba 0.48.0?

**Comment by qfwysw** [comment]
It's very strange, the version of numba is 0.48.0.

**Comment by Tai-Wang** [comment]
Can you directly import the module in the interactive session of python?

**Comment by qfwysw** [comment]
Yes, I can't import rotate_iou_gpu_eval in the interactive session of python. How can I check my compilation process？
![image](https://user-images.githubusercontent.com/53482524/116776539-784e9200-aa9b-11eb-89d4-9580f62ccc83.png)

**Comment by qfwysw** [comment]
Is it an error caused by my version reduction？

**Comment by Tai-Wang** [comment]
I am not sure whether they are directly related but if you downgrade mmcv and mmdet, I would recommend you delete the `build` folder and recompile mmdet3d.

**Comment by qfwysw** [comment]
I have tried, but it doesn't work. Do you have any environment configuration for cuda11.2？

**Comment by Wuziyi616** [comment]
I have not tried using MMDet3D on cuda11.2. It seems that your NVCC version is not the same as your torch's cuda version and the CUDA runtime version, which could probably cause the compilation error. I think you can try compiling torch, mmcv, mmdet and mmdet3d all from source because this will ensure the environment consistency?

**Comment by xiangruhuang** [comment]
I experienced similar issue with CUDA 11.2 days ago. It was resolved by using cuda 11.0 instead.

**Comment by THU17cyz** [comment]
Maybe simply numba 0.48.0 does not fully support CUDA 11.2. You may refer to the changelog of numba [here](https://pyup.io/changelogs/numba/#0.48.0).

**Comment by msmekala** [comment]
I have tried using MMDet3D on cuda 11.2, which causes the above error... There i have followed suggestions and i have experienced as bellow

Successfully uninstalled numba-0.50.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
mmdet3d 0.18.0 requires numba==0.48.0, but you have numba 0.53.0 which is incompatible.
Note: To solve this issue, It might resolved by using cuda 11.0 and less than 11.0 instead.

**Comment by msmekala** [comment]
> Thank you for your outstanding work！ **My Environment**
> 
> `fatal: not a git repository (or any parent up to mount point /) Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set). sys.platform: linux Python: 3.7.10 (default, Feb 26 2021, 18:47:35) [GCC 7.3.0] CUDA available: True GPU 0,1: GeForce RTX 3090 CUDA_HOME: /usr/local/cuda-11.2 NVCC: Build cuda_11.2.r11.2/compiler.29558016_0 GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0 PyTorch: 1.7.1+cu110 PyTorch compiling details: PyTorch built with:
> 
> * GCC 7.3
> * C++ Version: 201402
> * Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
> * Intel(R) MKL-DNN v1.6.0 (Git Hash 5ef631a030a6f73131c77892041042805a06064f)
> * OpenMP 201511 (a.k.a. OpenMP 4.5)
> * NNPACK is enabled
> * CPU capability usage: AVX2
> * CUDA Runtime 11.0
> * NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80
> * CuDNN 8.0.5
> * Magma 2.5.2
> * Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_VULKAN_WRAPPER -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,
> 
> TorchVision: 0.8.2+cu110 OpenCV: 4.5.1 MMCV: 1.2.7 MMCV Compiler: GCC 7.5 MMCV CUDA Compiler: 11.2 MMDetection: 2.10.0 MMDetection3D: 0.12.0+ `**The command I run**`python tools/train.py configs/pointpillars/hv_pointpillars_secfpn_6x8_160e_kitti-3d-3class.py`
> 
> **Describe the bug.** `"Import Error in mmdet3d.models.builder", I downgrade mmdet to 2.10.0 and downgrade mmcv to 1.2.7 to solve it. ` `When I start to train, the training looks good. After a while, there is a new error.` `2021-05-01 15:46:44,930 - mmdet - INFO - Saving checkpoint at 2 epochs [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 47.5 task/s, elapsed: 79s, ETA: 0s Converting prediction to KITTI format [>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 3769/3769, 451.3 task/s, elapsed: 8s, ETA: 0s Result is saved to /tmp/tmp13fakuid/results.pkl. Traceback (most recent call last): File "tools/train.py", line 212, in main() File "tools/train.py", line 208, in main meta=meta) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmdet/apis/train.py", line 170, in train_detector runner.run(data_loaders, cfg.workflow) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 125, in run epoch_runner(data_loaders[i], **kwargs) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmcv/runner/epoch_based_runner.py", line 54, in train self.call_hook('after_train_epoch') File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmcv/runner/base_runner.py", line 308, in call_hook getattr(hook, fn_name)(self) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmdet/core/evaluation/eval_hooks.py", line 147, in after_train_epoch key_score = self.evaluate(runner, results) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/mmdet/core/evaluation/eval_hooks.py", line 177, in evaluate results, logger=runner.logger, **self.eval_kwargs) File "/home/qf/mm/mmdetection3d/mmdet3d/datasets/kitti_dataset.py", line 355, in evaluate self.CLASSES) File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 707, in kitti_eval eval_types) File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 613, in do_eval min_overlaps) File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 479, in eval_class rets = calculate_iou_partly(dt_annos, gt_annos, metric, num_parts) File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 382, in calculate_iou_partly dt_boxes).astype(np.float64) File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/eval.py", line 116, in bev_box_overlap from .rotate_iou import rotate_iou_gpu_eval File "/home/qf/mm/mmdetection3d/mmdet3d/core/evaluation/kitti_utils/rotate_iou.py", line 292, in criterion=-1): File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/decorators.py", line 101, in kernel_jit kernel.bind() File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/compiler.py", line 548, in bind self._func.get() File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/compiler.py", line 426, in get ptx = self.ptx.get() File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/compiler.py", line 397, in get **self._extra_options) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 496, in llvm_to_ptx ptx = cu.compile(**opts) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 233, in compile self._try_error(err, 'Failed to compile\n') File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 251, in _try_error self.driver.check_error(err, "%s\n%s" % (msg, self.get_log())) File "/home/qf/conda/envs/open-mmlab/lib/python3.7/site-packages/numba/cuda/cudadrv/nvvm.py", line 141, in check_error raise exc numba.cuda.cudadrv.error.NvvmError: Failed to compile
> 
> (66, 23): parse expected comma after load's type NVVM_ERROR_COMPILATION `

Hi..... Have you resolved this issue?

**Comment by gbliao** [comment]
I experienced this issue with CUDA 11.6. It was resolved by using cuda 11.1 instead.

**Comment by ptrahuang** [comment]
cuda>=11.3
pip show numba
pip list --outdated --format=columns -i https://pypi.tuna.tsinghua.edu.cn/simple
pip install --upgrade numba -i https://pypi.tuna.tsinghua.edu.cn/simple
ok！
